libs/zedUpload/httputil/http.go

```diff
@@ -148,30 +150,112 @@ func ExecCmd(cmd, host, remoteFile, localFile string, objSize int64,
 			return stats
 		}
 		defer local.Close()
-		defer resp.Body.Close()
-		chunkSize := SingleMB
-		var written, copiedSize int64
-		stats.Size = objSize
-		for {
-			var copyErr error
-			if written, copyErr = io.CopyN(local, resp.Body, chunkSize); copyErr != nil && copyErr != io.EOF {
-				stats.Error = copyErr
-				return stats
+
+		var errorList []string
+		done := false
+		canSeek := false
+
+		delay := time.Second
+		lastModified := ""
+		var currentErrorList []string
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			if len(currentErrorList) > 0 {
+				// if we have previous errors, collect them
+				errorList = append(errorList, fmt.Sprintf("(attempt %d/%d):", attempt, maxRetries))
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				currentErrorList = []string{}
 			}
-			copiedSize += written
-			if written != chunkSize {
-				// Must have reached EOF
+			if ctx.Err() != nil {
+				errorList = append(errorList, fmt.Sprintf("ctx.Err: %v", ctx.Err()))
 				break
 			}
-			stats.Asize = copiedSize
-			if prgNotify != nil {
-				select {
-				case prgNotify <- stats:
-				default: //ignore we cannot write
+			if done {
+				break
+			}
+			if attempt > 0 {
+				time.Sleep(delay)
+				if delay < maxDelay {
+					delay = delay * 2
+				}
+			}
+			if canSeek && copiedSize > 0 {
+				req.Header.Set("Range", fmt.Sprintf("bytes=%d-", copiedSize))
+			}
+			resp, err := client.Do(req)
+			if err != nil {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("get failed: %s", err))
+				continue
+			}
+			if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("bad response code: %d", resp.StatusCode))
+				err = resp.Body.Close()
+				if err != nil {
+					currentErrorList = append(currentErrorList, fmt.Sprintf("error close Body: %v", err))
+				}
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				//we do not want to process server errors here
+				break
+			}
+			newLastModified := resp.Header.Get("Last-Modified")
+			if lastModified != "" && newLastModified != lastModified {
+				// last modified changed, retry without Range
+				lastModified = newLastModified
+				canSeek = false
+				continue
+			}
+			if resp.StatusCode == http.StatusOK {
+				stats.BodyLength = int(resp.ContentLength)
```

Comment:
I don't see any code which sets BodyLength to the total size when we get partial content. Isn't that needed?

Reply:
The first attempt is always without ranges, and we need only this size, not size of part from partial response

Reply:
ok, but the first one will fail, right? Due to a timeout or reset or something. So when does it get set?

Reply:
Got it that the first client.Do (without range) will return OK once and we will set BodyLength from that.